src/utils.js

```diff
@@ -0,0 +1,206 @@
+'use strict'
+
+const multihashing = require('multihashing-async')
+const CID = require('cids')
+const sort = require('stable').inplace
+const protobuf = require('protocol-buffers')
+const proto = protobuf(require('./dag.proto'))
+const DAGNode = require('./dag-node')
+const DAGLink = require('./dag-link')
+
+const hash = (type, data, cb) => multihashing(data, type, cb)
+
+const linkSort = (a, b) => {
+  return (new Buffer(a.name || '', 'ascii').compare(new Buffer(b.name || '', 'ascii')))
+}
+
+function cid (node, callback) {
+  callback(null, new CID(node.multihash))
+}
+
+function create (data, dagLinks, hashAlg, callback) {
+  if (typeof data === 'function') {
+    // empty obj
+    callback = data
+    data = undefined
+  }
+  if (typeof dagLinks === 'function') {
+    // empty obj
+    callback = dagLinks
+    dagLinks = []
+  }
+  if (typeof hashAlg === 'function') {
+    // empty obj
+    callback = hashAlg
+    hashAlg = undefined
+  }
+
+  if (!hashAlg) {
+    hashAlg = 'sha2-256'
+  }
+
+  const links = dagLinks.map((l) => {
+    if (!l.constructor && l.constructor.name !== 'DAGLink') {
+      return l
+    }
+
+    // haadcode: are the .name vs .Name for backwards compatibility?
+    const link = new DAGLink(l.name || l.Name,
+                             l.size || l.Size,
+                             l.hash || l.Hash || l.multihash)
+    return link
+  })
+
+  // Sort the links (in-place)
+  sort(links, linkSort)
+
+  serialize({
+    data: data,
+    links: links
+  }, (err, serialized) => {
+    if (err) {
+      callback(err)
+    }
+    multihashing(serialized, hashAlg, (err, multihash) => {
+      if (err) {
+        callback(err)
+      }
+      const dagNode = new DAGNode(data, links, serialized, multihash)
+      callback(null, dagNode)
+    })
+  })
+}
+
+function addLink (dagNode, nameOrLink, nodeOrMultihash, callback) {
+  const links = _cloneLinks(dagNode)
+  const data = _cloneData(dagNode)
+  let newLink = null
+
+  if ((nameOrLink.constructor &&
+       nameOrLink.constructor.name === 'DAGLink')) {
+    // It's a link
+    newLink = nameOrLink
+    // It's a name
+    if ((nodeOrMultihash.constructor &&
+       nodeOrMultihash.constructor.name === 'DAGNode')) {
+      // It's a node
+      newLink = toDAGLink(nodeOrMultihash)
+    } else {
+      // It's a multihash
+      newLink = new DAGLink(null, dagNode.size, nodeOrMultihash)
+    }
+  }
```

Comment:
This piece is incorrect, (also already gave reviews to this before). Again see line 67 to 97 https://github.com/ipld/js-ipld-dag-pb/pull/6#discussion_r88212711

Don't delete code + comments just because, information gets lost

Also, the addLink function, will need the `size` of link if just a name and multihash are passed.